x/ccv/provider/keeper/grpc_query.go

```diff
@@ -182,3 +182,32 @@ func (k Keeper) QueryProposedConsumerChainIDs(goCtx context.Context, req *types.
 		ProposedChains: chains,
 	}, nil
 }
+
+func (k Keeper) QueryAllPairsValConAddrByConsumerChainID(goCtx context.Context, req *types.QueryAllPairsValConAddrByConsumerChainIDRequest) (*types.QueryAllPairsValConAddrByConsumerChainIDResponse, error) {
+	if req == nil {
+		return nil, status.Error(codes.InvalidArgument, "empty request")
+	}
+
+	if req.ChainId == "" {
+		return nil, status.Error(codes.InvalidArgument, "empty chainId")
+	}
+
+	// list of pairs valconsensus addr <providerValConAddrs : consumerValConAddrs>
+	pairValConAddrs := []*types.PairValConAddrProviderAndConsumer{}
+
+	ctx := sdk.UnwrapSDKContext(goCtx)
+	validatorConsumerAddrs := k.GetAllValidatorsByConsumerAddr(ctx, &req.ChainId)
```

Comment:
I think that getting the consumer addresses by calling the `GetAllValidatorsByConsumerAddr` has an unintended behavior since 
```
// GetValidatorsByConsumerAddrs gets all the mappings from consensus addresses
// on a given consumer chain to consensus addresses on the provider chain.
```
This mapping contains also older entries that are pruned once no longer needed for slashing. For example, if a validator reassigns the key on a consumer, a new entry will be added to this mapping. From the description of the issue, these old entries shouldn't be return by the query. 

I think that `GetAllValidatorConsumerPubKeys` should be used instead. This function returns a list of tuples `(chainID, providerAddr, consumerKey)`. The consumer address can be obtained then from the key, i.e.,
```golang
consumerAddr, err := ccvtypes.TMCryptoPublicKeyToConsAddr(consumerKey)
```

Reply:
Make sense, I will update it now